engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/transformation/GenericNodeTransformation.scala

```diff
@@ -0,0 +1,62 @@
+package pl.touk.nussknacker.engine.api.context.transformation
+
+import pl.touk.nussknacker.engine.api.context.transformation.GenericNodeTransformation.NodeTransformationDefinition
+import pl.touk.nussknacker.engine.api.context.ValidationContext
+import pl.touk.nussknacker.engine.api.definition.{NodeDependency, Parameter, WithExplicitMethodToInvoke}
+import pl.touk.nussknacker.engine.api.typed.typing
+import pl.touk.nussknacker.engine.api.typed.typing.Unknown
+
+/*
+  The idea is this would be low-level API, with more accessible DSLs/builders - also with annotation based configuration
+  I'd like to handle only this trait (or sth similar) in compiler - instead of MethodToInvoke, WithExplicitMethodToInvoke etc
+  Note: this trait is used without @MethodToInvoke. Usages with @MethodToInvoke can be implemented using GenericNodeTransformation
+  by NK itself (using more or less current DefinitionExtractor logic)
+ */
+object GenericNodeTransformation {
+
+  //TODO: what if we cannot determine parameters/context? With some "fatal validation error"?
+  type NodeTransformationDefinition = PartialFunction[List[(String, DefinedParameter)], TransformationStepResult]
```

Comment:
WDYT about adding some State monad to it? I would like to add some extra information between steps that will be used in each step computation. Consider situation: you have initial context `a: { b: Int, c: Int }, d: Int` and want to have some extractions on each step that will cause context stripping - e.g. you want to put `b: Int, c: Int` to `a` param additional variables and only `d: Int` to `rest` param.

Reply:
I have a bit mixed feelings... I think it would be easier to ignore (if you don't need it) if you make it so:
PartialFunction[(List[(String, DefinedParameter), State)], TransformationStepResult]
but I wonder what State would be? I mean - we can make it some generic Map[String, Any] of course...

Reply:
`PartialFunction[(List[(String, DefinedParameter), State)], TransformationStepResult]` will be ok, but we will need to have some zero element for it then. For now I have some strange examples of `State` in mine use case. It was kind of `Optiona[ValidatationContext]` that at the beginning was `empty`, after initial step was `Some(fullListOfEnrichedAttributes)` and after extracting it to each parameter's it was reduced again to `Some(emptyListOfAttributes)`. I can imagine that it can be very different based on each case. I would totally prefer path dependent types way. For "normal users" we have high level API. Also someone can ignore state. and everything will looks pretty easy still IMO.

Reply:
Option[_] seems like good zero :) I added state with dependent types and it looks pretty neat I think